src/mlpack/core/data/dictionary_encoding_impl.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * @file dictionary_encoding_impl.hpp
+ * @author Jeffin Sam
+ *
+ * Implementation of dictionary encoding functions.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+#define MLPACK_CORE_DATA_DICT_ENCODING_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "dictionary_encoding.hpp"
+
+namespace mlpack {
+namespace data {
+
+void DicitonaryEncoding::Reset()
+{
+  mappings.clear();
+}
+
+template<typename TokenizerType>
+void DicitonaryEncoding::CreateMap(std::string& strings,
+    TokenizerType tokenizer)
+{
+  boost::string_view strView(strings);
+  boost::string_view token;
+  token = tokenizer(strView);
+  std::size_t curLabels = mappings.size() + 1;
+  while (!token.empty())
+  {
+    if (mappings.find(std::string(token)) == mappings.end())
```

Comment:
I think we could optimize this line. Actually, you create`std::string()` from each token. Thus, you get the same number of memory allocations as if you use `std::string`. I see 2 possible optimizations:
1) Replace `unordered_map<string, size_t>` by `unordered_map<string_view, size_t>`.
2) We could introduce a variable of type `std::string` outside of the loop. I think `operator=()` doesn't allocate memory if it has already been allocated. It could work if `operator=()` is defined for `std::string` and `boost::string_view`.

I am not sure which way is the best one. Maybe your current implementation works well. I suggest to write a simple benchmark.

Reply:
> I think we could optimize this line. Actually, you create`std::string()` from each token. Thus, you get the same number of memory allocations as if you use `std::string`. I see 2 possible optimizations:
> 
>     1. Replace `unordered_map<string, size_t>` by `unordered_map<string_view, size_t>`.
> 
>     2. We could introduce a variable of type `std::string` outside of the loop. I think `operator=()` doesn't allocate memory if it has already been allocated. It could work if `operator=()` is defined for `std::string` and `boost::string_view`.
> 
> 
> I am not sure which way is the best one. Maybe your current implementation works well. I suggest to write a simple benchmark.

Optimized the operation after we had discussion over irc, used `unordered_map<string_view, size_t>` along with the hasher, Thanks :) for the work around.